src/tarantool.client/LogicalConnection.cs

```diff
@@ -72,32 +137,36 @@ public static byte[] ReadFully(Stream input)
             }
         }
 
-        public IEnumerable<TaskCompletionSource<MemoryStream>> PopAllResponseCompletionSources()
-        {
-            var result = _pendingRequests.Values.ToArray();
-            _pendingRequests.Clear();
-            return result;
-        }
-
         private async Task<TResponse> SendRequestImpl<TRequest, TResponse>(TRequest request)
             where TRequest : IRequest
         {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(LogicalConnection));
+            }
+
             var bodyBuffer = MsgPackSerializer.Serialize(request, _msgPackContext);
 
-            var requestId = GetRequestId();
-            var responseTask = GetResponseTask(requestId);
+            var requestId = _requestIdCounter.GetRequestId();
+            var responseTask = _responseReader.GetResponseTask(requestId);
 
             long headerLength;
             var headerBuffer = CreateAndSerializeBuffer(request, requestId, bodyBuffer, out headerLength);
 
-            lock (_physicalConnection)
```

Comment:
This lock is necessary. Otherwise you'll end up with two threads mixing requests and broken pipeline